notebooks/05_Kubeflow_Pipeline/05_04_Pipeline_SageMaker.ipynb

```diff
@@ -68,9 +77,54 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "!aws s3 cp s3://kubeflow-pipeline-data/mnist_kmeans_example/data s3://$S3_BUCKET/mnist_kmeans_example/data\n",
+    "import pickle, gzip, numpy, urllib.request, json\n",
+    "from urllib.parse import urlparse\n",
+    "\n",
+    "# Load the dataset\n",
+    "urllib.request.urlretrieve(\"http://deeplearning.net/data/mnist/mnist.pkl.gz\", \"mnist.pkl.gz\")\n",
+    "with gzip.open('mnist.pkl.gz', 'rb') as f:\n",
+    "    train_set, valid_set, test_set = pickle.load(f, encoding='latin1')\n",
+    "\n",
+    "\n",
+    "# Upload dataset to S3\n",
+    "from sagemaker.amazon.common import write_numpy_to_dense_tensor\n",
+    "import io\n",
+    "import boto3\n",
     "\n",
-    "!aws s3 cp s3://kubeflow-pipeline-data/mnist_kmeans_example/input/valid_data.csv s3://$S3_BUCKET/mnist_kmeans_example/input/"
+    "###################################################################\n",
+    "# This is the only thing that you need to change to run this code \n",
+    "# Give the name of your S3 bucket \n",
+    "bucket = S3_BUCKET \n",
+    "\n",
+    "# If you are gonna use the default values of the pipeline then \n",
+    "# give a bucket name which is in us-west-2 region \n",
+    "###################################################################\n",
+    "\n",
+    "train_data_key = 'mnist_kmeans_example/train_data'\n",
+    "test_data_key = 'mnist_kmeans_example/test_data'\n",
+    "train_data_location = 's3://{}/{}'.format(bucket, train_data_key)\n",
+    "test_data_location = 's3://{}/{}'.format(bucket, test_data_key)\n",
+    "print('training data will be uploaded to: {}'.format(train_data_location))\n",
+    "print('training data will be uploaded to: {}'.format(test_data_location))\n",
+    "\n",
+    "# Convert the training data into the format required by the SageMaker KMeans algorithm\n",
+    "buf = io.BytesIO()\n",
+    "write_numpy_to_dense_tensor(buf, train_set[0], train_set[1])\n",
+    "buf.seek(0)\n",
+    "\n",
+    "boto3.resource('s3').Bucket(bucket).Object(train_data_key).upload_fileobj(buf)\n",
+    "\n",
+    "# Convert the test data into the format required by the SageMaker KMeans algorithm\n",
+    "write_numpy_to_dense_tensor(buf, test_set[0], test_set[1])\n",
+    "buf.seek(0)\n",
+    "\n",
+    "boto3.resource('s3').Bucket(bucket).Object(test_data_key).upload_fileobj(buf)\n",
+    "\n",
+    "# Convert the valid data into the format required by the SageMaker KMeans algorithm\n",
+    "numpy.savetxt('valid-data.csv', valid_set[0], delimiter=',', fmt='%g')\n",
+    "s3_client = boto3.client('s3')\n",
```

Comment:
Can you modify validation data conversion to be the same pattern as you did in the training and test data conversion?

Reply:
the validate-data.csv file is a different format.